### PR TITLE
feat: Bulk import products from Shopify

### LINKED ISSUE #59: Bulk import of catalog from shopify
- Bulk import of catalogue without receipt of any order.

- Good to have: configurable naming series (adds complexity)

### PR DESCRIPTION
**Pull request by penieldev**
Import products from shopify without waiting for orders. 

image

[no-docs]

Resolves #59

### CODE REVIEW

**Comment by ankush** [comment]
@penieldev awesome contribution 🙌 Please bear with me if it takes a somewhat longer time for review because this kinda huge contribution and requires some manual testing too.

File: ecommerce_integrations/public/build.json
Comment (ankush) [inline ecommerce_integrations/public/build.json:3]: Need to confirm if this will work well with new build system.

Comment (ankush) [inline ecommerce_integrations/public/build.json:3]: note to self: need to make sure if this will work with new build system.

**Comment by penieldev** [comment]
> @penieldev awesome contribution 🙌 Please bear with me if it takes a somewhat longer time for review because this kinda huge contribution and requires some manual testing too.
> 
> It would be really helpful if you add a simple test for bulk import functionality (not UI, just python side code)
> 
> https://github.com/frappe/ecommerce_integrations/blob/develop/ecommerce_integrations/shopify/tests/test_product.py https://github.com/frappe/ecommerce_integrations/blob/develop/ecommerce_integrations/shopify/tests/utils.py
> 
> These files have examples to fake shopify API response.

No problem. Take your time. I need time as well to figure out the tests. I'll push it here when I have them working.

**Comment by penieldev** [comment]
image

I had the same issue locally, and I had to add those ledgers in my CoA to fix it. What should I do here? 

PS: Sorry for taking so long to respond.

**Comment by ankush** [comment]
IIRC I fixed that issue in `develop` branch, can you merge develop into your branch? @penieldev

**Comment by ankush** [comment]
@penieldev hi you'll have to merge again because latest frappe version wasn't installable in CI due to python version mismatch 😅 

Can you give maintainers edit rights, so I can push small changes like this myself?

image

**Comment by penieldev** [comment]
> @penieldev hi you'll have to merge again because latest frappe version wasn't installable in CI due to python version mismatch 😅
> 
> Can you give maintainers edit rights, so I can push small changes like this myself?
> 
> image

I do not have that checkbox since it is an organization repo. I can add you as a maintainer to the repo itself if you want.

**Comment by penieldev** [comment]
Is there any way I can run these actions locally so that I won't use up the Action Minutes?

**Comment by lucasvazq** [comment]
Hello @penieldev if you are talking about Github Actions, there is a tool that allows you to run them locally.

https://github.com/nektos/act

**Comment by penieldev** [comment]
@lucasvazq Thanks. I will check it out. I've managed to install **pre-commit** and **semgrep** in the virtualenv itself. It gets the job done for now.

File: ecommerce_integrations/public/build.json
Comment (ankush) [reply ecommerce_integrations/public/build.json:3]: Screenshot 2022-01-12 at 7 55 43 PM


Doesn't work with new build system. I am guessing you use version-13 branch.

Comment (ankush) [reply ecommerce_integrations/public/build.json:3]: Here's the PR that added new build system.

IMO we can also just move all JS to page.js instead of bundling it. It will reduce complexity a lot.

Comment (ankush) [reply ecommerce_integrations/public/build.json:3]: This page would be used occasionally, so not much benefit from bundling the files.

Comment (penieldev) [reply ecommerce_integrations/public/build.json:3]: > This page would be used occasionally, so not much benefit from bundling the files.

Yes. Makes sense. The idea was to create a base class that can be inherited to import products, customers, and other stuff. But now I don't know if it's worth it.

**Comment by penieldev** [comment]
What would be the ideal location to add a button or link to this page? Currently the user will have to use the awesomebar to find the page. 

Also, is 'Import Shopify Products' the right name for this page? I feel like changing the name would be a lot harder after merging.

**Comment by ankush** [comment]
@penieldev 

Linking it as a button Shopify setting page should be good enough for now. Currently, there's no workspace link for this app. 

re: name. It's hard to name things 🥲 but the name you've chosen conveys exactly what it does, so it's fine IMO.

File: ecommerce_integrations/shopify/page/shopify_import_products/shopify_import_products.py
Comment (ankush) [inline ecommerce_integrations/shopify/page/shopify_import_products/shopify_import_products.py:89]:
```suggestion
	except Exception:
		frappe.db.rollback()
		return False
```

If you catch exception then framework won't know that something has gone wrong during handling of a request so we need to rollback manually. 

https://frappeframework.com/docs/user/en/api/database#database-transaction-model

File: ecommerce_integrations/shopify/page/shopify_import_products/shopify_import_products.py
Comment (ankush) [inline ecommerce_integrations/shopify/page/shopify_import_products/shopify_import_products.py:129]: Same issue about transaction here, I'd recommend using savepoints to partially rollback till the last product sync. 


example here:  https://github.com/frappe/frappe/pull/15379

File: ecommerce_integrations/shopify/page/shopify_import_products/shopify_import_products.py
Comment (ankush) [inline ecommerce_integrations/shopify/page/shopify_import_products/shopify_import_products.py:139]:
```suggestion
			frappe.db.commit()  # prevents too many write request error
```

Simple commit here should be sufficient. IDK any reason why commit will fail 😅

**Review by ankush — CHANGES_REQUESTED** [review]
some small changes related to exception handling, rest looks good 👍

**Review by ankush — APPROVED** [review]
LGTM; let me know if you need to change anything or will merge in 2ish  days.

**Comment by penieldev** [comment]
Just need to add the button and we're good to go.

File: ecommerce_integrations/shopify/doctype/shopify_setting/shopify_setting.js
Comment (ankush) [inline ecommerce_integrations/shopify/doctype/shopify_setting/shopify_setting.js:31]:
```suggestion
		frm.add_custom_button(__('Import Products'), function () {
```

translation

**Pull request merged**
